promoter/promoter.go

```diff
@@ -186,13 +193,90 @@ func (p *Promoter) staticAddrDiff(ctx context.Context) (toAdd []WatchedAddress,
 // threadedCreditTransactions continuously polls the db for uncreditted txns and
 // notifies the credit promoter about them.
 func (p *Promoter) threadedCreditTransactions() {
-	// TODO: Fetch all txns with credited == false.
+	t := time.NewTicker(txnPollInterval)
+	defer t.Stop()
+LOOP:
+	for {
+		select {
+		case <-p.staticBGCtx.Done():
+			return
+		case <-t.C:
+		}
+
+		// Loop over txns one-by-one.
+		for {
+			// Fetch an transaction that the credit system doesn't know
+			// about yet.
+			currentTime := time.Now().UTC()
+			sr := p.staticColTransactions().FindOneAndUpdate(p.staticBGCtx, bson.M{
+				"credited_at": bson.M{
+					"$lt": currentTime.Add(-txnPollInterval),
+				},
+			}, bson.M{
+				"$set": bson.M{
+					"credited_at": currentTime,
+				},
+			}, options.FindOneAndUpdate())
+			if errors.Contains(sr.Err(), mongo.ErrNoDocuments) {
+				continue LOOP // no more txns in this iteration
+			}
+			if sr.Err() != nil {
+				p.staticLogger.WithError(sr.Err()).Error("Failed to fetch another uncredited txn")
+				continue LOOP // db failure, try again later
+			}
+
+			// Decode txn.
+			var txn Transaction
+			if err := sr.Decode(&txn); err != nil {
+				p.staticLogger.WithError(err).Error("Failed to decode txn")
+				continue // try next txn
+			}
+
+			// Fetch the user for the txn.
+			sr = p.staticColWatchedAddresses().FindOne(p.staticBGCtx, bson.M{
+				"_id": txn.Address,
+			})
+			if errors.Contains(sr.Err(), mongo.ErrNoDocuments) {
+				p.staticLogger.WithError(sr.Err()).Error("Address for txn doesn't exist")
+				continue // try next
+			}
```

Comment:
nit: If we get a transaction that we can't match to a user, we should probably alert the operator. `build.Critical`? The same might be true for the ones we fail to decode.

Should we mark these with the number of times we fail to process them and eventually give up? If not, we should definitely document in the README that the operator needs to watch for these errors.